### PR TITLE
refactor signum30

### PR DESCRIPTION
**Pull request by sgurenkov**
## Description

Refactor `signum` keyboard to allow different microcontrollers (teensy)
Add one more keymap for signum/3_0

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

cc: @jceb

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/signum/3_0/config.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#pragma once
```
Comment (drashna) [inline keyboards/signum/3_0/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/signum/3_0/config.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0x6060
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    "Troy Fletcher"
```
Comment (drashna) [inline keyboards/signum/3_0/config.h:9]:
```suggestion
#define MANUFACTURER    Troy Fletcher
```

File: keyboards/signum/3_0/keymaps/sgurenkov/config.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#pragma once
```
Comment (drashna) [inline keyboards/signum/3_0/keymaps/sgurenkov/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/signum/3_0/keymaps/sgurenkov/keymap.c [hunk 4]
```diff
@@ -0,0 +1,140 @@
+#include "signum.h"
```
Comment (drashna) [inline keyboards/signum/3_0/keymaps/sgurenkov/keymap.c:1]:
```suggestion
#include QMK_KEYBOARD_H
```

Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/signum/3_0/signum.c [hunk 5]
```diff
@@ -0,0 +1 @@
+#include "signum.h"
```
Comment (drashna) [inline keyboards/signum/3_0/signum.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/signum/3_0/signum.h [hunk 6]
```diff
@@ -0,0 +1,17 @@
+#pragma once
```
Comment (drashna) [inline keyboards/signum/3_0/signum.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/signum/3_0/teensy/config.h [hunk 7]
```diff
@@ -0,0 +1,6 @@
+#pragma once
```
Comment (drashna) [inline keyboards/signum/3_0/teensy/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/signum/3_0/teensy/teensy.h [hunk 8]
```diff
@@ -0,0 +1,4 @@
+#pragma once
```
Comment (drashna) [inline keyboards/signum/3_0/teensy/teensy.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/signum/3_0/teensy/teensy.c [hunk 9]
```diff
@@ -0,0 +1 @@
+#include "teensy.h"
```
Comment (drashna) [inline keyboards/signum/3_0/teensy/teensy.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/signum/3_0/rules.mk [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+EXTRAKEY_ENABLE = yes	# Audio control and System control
+CONSOLE_ENABLE = no	# Console for debug
+COMMAND_ENABLE = no    # Commands for debug and configuration
+SLEEP_LED_ENABLE = no  # Breathing sleep LED during USB suspend
+NKRO_ENABLE = yes		# USB Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no  # Enable keyboard backlight functionality
+AUDIO_ENABLE = no
+RGBLIGHT_ENABLE = no
+
+UNICODEMAP_ENABLE = no
```
Comment (fauxpark) [inline keyboards/signum/3_0/rules.mk:21]:
```suggestion
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/signum/3_0/keymaps/sgurenkov/keymap.c [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+LEADER_EXTERNS();
+
+void matrix_scan_user(void) {
+  LEADER_DICTIONARY() {
+    leading = false;
+    leader_end();
+
+    // Cancel task CTRL+C
+    SEQ_ONE_KEY(KC_C) {
+      SEND_STRING(SS_LCTL("c"));
```
Comment (fauxpark) [inline keyboards/signum/3_0/keymaps/sgurenkov/keymap.c:91]: These can be simplified:
```suggestion
      tap_code16(C(KC_C));
```

File: keyboards/signum/3_0/keymaps/sgurenkov/keymap.c [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+      unregister_code(KC_BSPACE);
+
+      unregister_code(KC_LGUI);
+    }
+    // go to the beginning of the string
+    SEQ_ONE_KEY(KC_H) {
+      register_code(KC_LCMD);
+      register_code(KC_LEFT);
+      unregister_code(KC_LEFT);
+      unregister_code(KC_LCMD);
```
Comment (fauxpark) [inline keyboards/signum/3_0/keymaps/sgurenkov/keymap.c:130]:
```suggestion
      tap_code16(G(KC_LEFT));
```

File: keyboards/signum/3_0/keymaps/sgurenkov/keymap.c [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    // go to the end of the string
+    SEQ_ONE_KEY(KC_L) {
+      register_code(KC_LCMD);
+      register_code(KC_RIGHT);
+      unregister_code(KC_RIGHT);
+      unregister_code(KC_LCMD);
```
Comment (fauxpark) [inline keyboards/signum/3_0/keymaps/sgurenkov/keymap.c:137]:
```suggestion
      tap_code16(G(KC_RIGHT));
```

File: keyboards/signum/3_0/keymaps/sgurenkov/keymap.c [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    // delete line
+    SEQ_TWO_KEYS(KC_D, KC_D) {
+      register_code(KC_LGUI);
+
+      register_code(KC_RIGHT);
+      unregister_code(KC_RIGHT);
+
+      register_code(KC_BSPACE);
+      unregister_code(KC_BSPACE);
```
Comment (fauxpark) [inline keyboards/signum/3_0/keymaps/sgurenkov/keymap.c:121]:
```suggestion
      tap_code(KC_RIGHT);

      tap_code(KC_BSPACE);
```

File: keyboards/signum/3_0/keymaps/sgurenkov/keymap.c [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    leading = false;
+    leader_end();
+
+    // Cancel task CTRL+C
+    SEQ_ONE_KEY(KC_C) {
+      SEND_STRING(SS_LCTL("c"));
+    }
+    // copy
+    SEQ_ONE_KEY(KC_Y) {
+      SEND_STRING(SS_LCMD("c"));
```
Comment (fauxpark) [inline keyboards/signum/3_0/keymaps/sgurenkov/keymap.c:95]:
```suggestion
      tap_code16(G(KC_C));
```

File: keyboards/signum/3_0/keymaps/sgurenkov/keymap.c [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    // cut
+    SEQ_ONE_KEY(KC_X) {
+      SEND_STRING(SS_LCMD("x"));
```
Comment (fauxpark) [inline keyboards/signum/3_0/keymaps/sgurenkov/keymap.c:99]:
```suggestion
      tap_code16(G(KC_X));
```

File: keyboards/signum/3_0/keymaps/sgurenkov/keymap.c [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    // paste
+    SEQ_ONE_KEY(KC_P) {
+      SEND_STRING(SS_LCMD("v"));
```
Comment (fauxpark) [inline keyboards/signum/3_0/keymaps/sgurenkov/keymap.c:103]:
```suggestion
      tap_code16(G(KC_V));
```

File: keyboards/signum/3_0/keymaps/sgurenkov/keymap.c [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    // undo
+    SEQ_ONE_KEY(KC_U) {
+      SEND_STRING(SS_LCMD("z"));
```
Comment (fauxpark) [inline keyboards/signum/3_0/keymaps/sgurenkov/keymap.c:107]:
```suggestion
      tap_code16(G(KC_Z));
```

File: keyboards/signum/3_0/keymaps/sgurenkov/keymap.c [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    // redo
+    SEQ_ONE_KEY(KC_R) {
+      SEND_STRING(SS_LCMD(SS_LSFT("z")));
```
Comment (fauxpark) [inline keyboards/signum/3_0/keymaps/sgurenkov/keymap.c:111]:
```suggestion
      tap_code16(S(G(KC_Z)));
```

File: keyboards/signum/3_0/keymaps/sgurenkov/keymap.c [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______),
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  switch (keycode) {
+    case CMD_TAB:
+        if (record->event.pressed) {
+            register_code(KC_LCMD);
+            register_code(KC_TAB);
+            unregister_code(KC_TAB);
```
Comment (fauxpark) [inline keyboards/signum/3_0/keymaps/sgurenkov/keymap.c:69]:
```suggestion
            tap_code(KC_TAB);
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by sgurenkov** [comment]
@drashna @fauxpark Thank you for the review and suggestions!
I've updated the PR, please give it another look

Thank you

File: keyboards/signum/3_0/readme.md [hunk 21]
```diff
@@ -9,6 +9,7 @@ Hardware available here: [Troy Fletcher](http://troyfletcher.net/keyboard_sales.
 
 Make example for this keyboard (after setting up your build environment):
 
-    make signum/3_0/elitec:default
+   make signum/3_0/elitec:default
+   make signum/3_0/teensy:default
```
Comment (fauxpark) [inline keyboards/signum/3_0/readme.md:13]:
```suggestion
    make signum/3_0/elitec:default
    make signum/3_0/teensy:default
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
One last thing.

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Pull request merged**